### PR TITLE
Make workflow task recovery workflow-aware

### PR DESCRIPTION
**Pull request by lsm**
Routes workflow-backed Reopen/Resume through runtime recovery so task, workflow run, and node execution state are restored together. Also clears stale completed timestamps when task/run state becomes active and labels workflow recovery actions in the task pane.

Tests cover cancelled workflow resume, task/run consistency, live vs stale node execution recovery, unchanged plain-task behavior, RPC routing, and web store/pane behavior.

### CODE REVIEW

**Review by lsm — COMMENTED** [review]
## 🤖 Review by Claude Opus 4.7 (Anthropic)

> **Model:** Claude Opus 4.7 | **Client:** NeoKai | **Provider:** Anthropic

### Verdict: REQUEST_CHANGES

Solid implementation overall — the snap-back prevention architecture is correct, the SQLite transaction ensures atomicity, and the executor registration after recovery is well-sequenced. The UI labels and routing logic are clean. That said, there are test gaps and a couple of robustness items to address.

---

### P2 — Missing runtime test for `done + workflowRunId → in_progress`

**`packages/daemon/tests/unit/5-space/runtime/space-runtime.test.ts`**

The "Reopen workflow" action (`done → in_progress`) is the second most common recovery path after `cancelled → in_progress`. The runtime test suite only covers the `cancelled` source status. While the code path through `recoverWorkflowBackedTask` is structurally identical regardless of source, the `done → in_progress` run transition (`VALID_TRANSITIONS.done → in_progress`) and the task transition (`done: ['in_progress', 'archived']`) are distinct from the `cancelled` case and warrant explicit coverage — especially verifying that the run's `completedAt` is properly cleared (since `done` runs always have `completedAt` set).

### P2 — Missing runtime test for `blocked + workflowRunId → open/in_progress`

**`packages/daemon/tests/unit/5-space/runtime/space-runtime.test.ts`**

The `blocked → open` case is particularly interesting because the task goes to `open` while the run always transitions to `in_progress` — an asymmetric state that should be verified to work correctly with the tick loop (i.e., the run is processed without snapping the task back). The `blocked → in_progress` case is also untested.

### P2 — Missing test for node execution recreation (empty executions path)

**`packages/daemon/src/lib/space/runtime/space-runtime.ts` lines 1172–1192**

The defensive branch that recreates start-node executions when `listByWorkflowRun` returns an empty array is completely untested. This is an edge case but it contains non-trivial logic (workflow lookup, start-node resolution, `resolveNodeAgents` iteration, `createOrIgnore`).

### P2 — `prepareSubSessionForWorkflowResume` return value silently ignored

**`packages/daemon/src/lib/space/runtime/space-runtime.ts` line 1234–1235**

```ts
for (const sessionId of liveSessionIds) {
    await this.config.taskAgentManager?.prepareSubSessionForWorkflowResume(sessionId);
}
```

If MCP self-heal fails (returns `false`), the session remains alive but missing `node-agent` and `space-agent-tools` MCP servers. The agent would be functional but unable to use workflow tools, leading to confusing errors. Consider at minimum logging a warning when the return is `false`.

### P3 — `isWorkflowRecoveryTransition` is broader than `VALID_SPACE_TASK_TRANSITIONS`

**`packages/daemon/src/lib/rpc-handlers/space-task-handlers.ts` line 24–32, `packages/web/src/components/space/SpaceTaskPane.tsx` line 88–96**

The function includes `done → open`, but `VALID_SPACE_TASK_TRANSITIONS['done']` is `['in_progress', 'archived']` — so the UI never generates this transition, and the recovery method rejects it via `isValidSpaceTaskTransition`. The function is harmless but misleading. Consider removing `done → open` from the check, or adding a comment explaining the intentional breadth.

### P3 — `isWorkflowRecoveryTransition` duplicated across daemon and web

**`packages/daemon/src/lib/rpc-handlers/space-task-handlers.ts` and `packages/web/src/components/space/SpaceTaskPane.tsx`**

The same function exists in both packages. If the transition rules change, both must be updated in sync. Consider moving to `@neokai/shared` to single-source the definition.

---

### What's done well

- **Snap-back prevention is correct**: The run is transitioned to `in_progress` and the executor is registered in the same logical operation, so the tick loop never sees an inconsistent state.
- **Atomicity**: All DB mutations (task, run, node executions) run in a single SQLite transaction.
- **Timestamp fix**: The repository-level auto-clearing of `completedAt` on transitions to `in_progress`/`open` (with explicit `completedAt` opt-out) is a clean defense-in-depth fix for the `startedAt > completedAt` buglet.
- **Node execution handling**: Live sessions are preserved and re-prepared; stale sessions are fully reset. The `prepareSubSessionForWorkflowResume` approach of re-running MCP self-heal is elegant.
- **Non-workflow tasks unaffected**: The recovery path is only entered when `workflowRunId` is truthy and the transition matches, so plain tasks go through the existing `setTaskStatus` path unchanged.
- **UI labels**: `getTaskActionLabel` provides clear "Reopen workflow" / "Resume workflow" labels that help users understand they're resuming the entire workflow, not just the task.

File: packages/daemon/tests/unit/5-space/runtime/space-runtime.test.ts [hunk 1]
```diff
@@ -192,6 +193,135 @@ describe('SpaceRuntime', () => {
 	// getRulesForStep — removed in M71; no tests needed
 	// -------------------------------------------------------------------------
 
+	describe('recoverWorkflowBackedTask()', () => {
```
Comment (lsm) [inline packages/daemon/tests/unit/5-space/runtime/space-runtime.test.ts:196]: **P2** — The `done → in_progress` recovery path ("Reopen workflow") is untested at the runtime level. This is the second most common recovery action. Consider adding a test analogous to the existing `cancelled → in_progress` tick test, verifying the run transitions from `done` to `in_progress` and `completedAt` is cleared.

File: packages/daemon/tests/unit/5-space/runtime/space-runtime.test.ts [hunk 2]
```diff
@@ -192,6 +193,135 @@ describe('SpaceRuntime', () => {
...
+			expect(reopened.workflowRunId).toBeUndefined();
+			expect(reopened.completedAt).toBeNull();
+
+			taskRepo.updateTask(task.id, { status: 'cancelled', completedAt });
+			const resumed = await manager.setTaskStatus(task.id, 'in_progress');
+			expect(resumed.status).toBe('in_progress');
+			expect(resumed.workflowRunId).toBeUndefined();
+			expect(resumed.completedAt).toBeNull();
+		});
+	});
```
Comment (lsm) [inline packages/daemon/tests/unit/5-space/runtime/space-runtime.test.ts:468]: **P2** — The `blocked → open` recovery case is particularly interesting because the task goes to `open` while the run always transitions to `in_progress` (line 1145). This asymmetric state should be verified to work correctly with the tick loop — the run should be processed without snapping the task back to `blocked`.

File: packages/daemon/src/lib/space/runtime/space-runtime.ts [hunk 3]
```diff
@@ -1100,6 +1103,148 @@ export class SpaceRuntime {
...
+				pendingCompletionReason: null,
+				postApprovalSessionId: null,
+				postApprovalStartedAt: null,
+				postApprovalBlockedReason: null,
+				reportedStatus: null,
+				reportedSummary: null,
+			});
+			if (!updatedTask) throw new Error(`Failed to update task: ${task.id}`);
+
+			let executions = this.config.nodeExecutionRepo.listByWorkflowRun(run.id);
```
Comment (lsm) [inline packages/daemon/src/lib/space/runtime/space-runtime.ts:1172]: **P2** — This defensive branch (recreating start-node executions when none exist) contains non-trivial logic (workflow lookup, start-node resolution, `resolveNodeAgents` iteration, `createOrIgnore`) but has no test coverage. Consider adding a test that deletes all node executions for a run and then calls recovery.

File: packages/daemon/src/lib/space/runtime/space-runtime.ts [hunk 4]
```diff
@@ -1100,6 +1103,148 @@ export class SpaceRuntime {
...
+					});
+				}
+			}
+
+			return { task: updatedTask, run: updatedRun };
+		});
+
+		const recovered = recoverTx();
+		await this.ensureExecutorRegistered(recovered.run);
+		for (const sessionId of liveSessionIds) {
```
Comment (lsm) [inline packages/daemon/src/lib/space/runtime/space-runtime.ts:1234]: **P2** — If `prepareSubSessionForWorkflowResume` returns `false` (MCP self-heal failed), the session remains alive but missing `node-agent` and `space-agent-tools` MCP servers. The agent would be functional but unable to use workflow tools, leading to confusing downstream errors. Consider logging a warning when the return is `false`.

File: packages/daemon/src/lib/rpc-handlers/space-task-handlers.ts [hunk 5]
```diff
@@ -17,6 +23,16 @@ import type { SpaceRuntimeService } from '../space/runtime/space-runtime-service
 
 const log = new Logger('space-task-handlers');
```
Comment (lsm) [inline packages/daemon/src/lib/rpc-handlers/space-task-handlers.ts:24]: **P3** — `done → open` is included here but `VALID_SPACE_TASK_TRANSITIONS['done']` is `['in_progress', 'archived']`. The UI never generates `done → open`, and `recoverWorkflowBackedTask` rejects it via `isValidSpaceTaskTransition`. This is harmless but misleading. Consider removing it or adding a comment.

File: packages/web/src/components/space/SpaceTaskPane.tsx [hunk 6]
```diff
@@ -85,6 +85,30 @@ const PRIORITY_BADGE_CLASSES: Record<SpaceTaskPriority, string> = {
 	urgent: 'border-red-500/30 bg-red-500/10 text-red-300',
 };
 
+function isWorkflowRecoveryTransition(
```
Comment (lsm) [inline packages/web/src/components/space/SpaceTaskPane.tsx]: **P3** — This function is duplicated in `SpaceTaskPane.tsx` (line 88). If the transition rules change, both locations must be updated. Consider moving to `@neokai/shared`.

File: packages/daemon/tests/unit/5-space/runtime/space-runtime.test.ts (hunk 1 above)
Comment (lsm) [reply packages/daemon/tests/unit/5-space/runtime/space-runtime.test.ts:196]: Added a runtime test for done -> in_progress workflow recovery. It verifies the task and run return to in_progress, stale task result is cleared, and run completedAt is cleared before the tick continues.

File: packages/daemon/tests/unit/5-space/runtime/space-runtime.test.ts (hunk 2 above)
Comment (lsm) [reply packages/daemon/tests/unit/5-space/runtime/space-runtime.test.ts:468]: Added blocked workflow recovery coverage for both blocked -> open and blocked -> in_progress. The open case verifies the asymmetric task=open/run=in_progress state survives a runtime tick without snapping back, and both cases verify completedAt/blocking metadata is cleared.

File: packages/daemon/src/lib/space/runtime/space-runtime.ts (hunk 3 above)
Comment (lsm) [reply packages/daemon/src/lib/space/runtime/space-runtime.ts:1172]: Added runtime coverage for the empty node-executions recovery branch. The test deletes all executions for a cancelled run, recovers it, and verifies start-node executions are recreated for each start-node agent in pending state.

File: packages/daemon/src/lib/rpc-handlers/space-task-handlers.ts (hunk 5 above)
Comment (lsm) [reply packages/daemon/src/lib/rpc-handlers/space-task-handlers.ts:24]: Handled by moving the predicate to shared code and narrowing it to valid recovery transitions. done -> open is no longer included; done only routes workflow recovery for done -> in_progress.

File: packages/daemon/src/lib/space/runtime/space-runtime.ts (hunk 4 above)
Comment (lsm) [reply packages/daemon/src/lib/space/runtime/space-runtime.ts:1234]: Handled. recoverWorkflowBackedTask now checks prepareSubSessionForWorkflowResume return values and logs a warning when a live session could not be prepared with workflow MCP tools.

File: packages/web/src/components/space/SpaceTaskPane.tsx (hunk 6 above)
Comment (lsm) [reply packages/web/src/components/space/SpaceTaskPane.tsx]: Handled. isWorkflowRecoveryTransition now lives in @neokai/shared and both the daemon RPC handler and SpaceTaskPane import the shared predicate.

**Review by lsm — COMMENTED** [review]
## 🤖 Review by Claude Opus 4.7 (Anthropic)

> **Model:** Claude Opus 4.7 | **Client:** NeoKai | **Provider:** Anthropic

### Verdict: APPROVE

All six findings from round 1 have been addressed:

- **P2 fixes**: Added runtime tests for `done→in_progress`, `blocked→open`, `blocked→in_progress` recovery paths, plus a test for the defensive node execution recreation branch. The `prepareSubSessionForWorkflowResume` return value is now checked with a warning log on failure.
- **P3 fixes**: `isWorkflowRecoveryTransition` moved to `@neokai/shared` (single source of truth) and narrowed to match `VALID_SPACE_TASK_TRANSITIONS` exactly (removed unreachable `done→open`).

The implementation is solid — snap-back prevention, transaction atomicity, executor registration, and UI routing are all correct.
